src/folding/hypernova/nimfs.rs

```diff
@@ -0,0 +1,657 @@
+use ark_ec::CurveGroup;
+use ark_ff::{Field, PrimeField};
+use ark_std::{One, Zero};
+
+use espresso_subroutines::PolyIOP;
+use espresso_transcript::IOPTranscript;
+
+use super::cccs::{Witness, CCCS};
+use super::lcccs::LCCCS;
+use super::utils::{compute_c_from_sigmas_and_thetas, compute_g, compute_sigmas_and_thetas};
+use crate::ccs::CCS;
+use crate::utils::hypercube::BooleanHypercube;
+use crate::utils::sum_check::structs::IOPProof as SumCheckProof;
+use crate::utils::sum_check::{verifier::interpolate_uni_poly, SumCheck};
+use crate::utils::virtual_polynomial::VPAuxInfo;
+
+use std::marker::PhantomData;
+
+/// Proof defines a multifolding proof
+#[derive(Debug)]
+pub struct Proof<C: CurveGroup> {
+    pub sc_proof: SumCheckProof<C::ScalarField>,
+    pub sigmas_thetas: SigmasThetas<C::ScalarField>,
+}
+
+#[derive(Debug)]
+pub struct SigmasThetas<F: PrimeField>(pub Vec<Vec<F>>, pub Vec<Vec<F>>);
+
+#[derive(Debug)]
+/// Implements the Non-Interactive Multi Folding Scheme described in section 5 of
+/// [HyperNova](https://eprint.iacr.org/2023/573.pdf)
+pub struct NIMFS<C: CurveGroup> {
+    pub _c: PhantomData<C>,
+}
```

Comment:
Still think that maybe inside of the NIMFS we should store the latest LCCCS instance that was folded in the previous fold. As well as `v`s, `u`s and `x`s and the rng.

Reply:
The idea is that the `NIMFS` logic supports folding multiple LCCCS instances with multiple CCCS instances ([eg](
https://github.com/privacy-scaling-explorations/folding-schemes/tree/b2c2f3fb7b24c4ec75d186b036aecdb6fc21ec23/src/folding/hypernova/nimfs.rs#L38)), and if we put the LCCCS as a single running instance then we can not support that.
Later, at the [`FoldingScheme`](
https://github.com/privacy-scaling-explorations/folding-schemes/tree/b2c2f3fb7b24c4ec75d186b036aecdb6fc21ec23/src/lib.rs#L53) trait level, when implementing it for HyperNova we will restrict it to a single LCCCS instance as the single 'running_instance', but at the NIMFS level I was trying to keep it more close to the scheme proposed in the HyperNova paper (with being able to fold multiple LCCCS & multiple CCCS in a single fold).

Reply:
I don't agree.
You always fold N->1 in HyperNova. Hence, at the end of each folding step you'll end up with a single LCCCS instance as a result of the folding. (Hence why NIMFS should only have 1 LCCS instance). Later, you can fold again with multiple LCCCS (other NIMFS accumulators) & CCCS instances.  

In any case, not that we're going to deal with this now anyways. Just wanted to at least point it for future discussion.

Reply:
I'm not sure we're referring to the same object/logic. For the IVC I agree that we will have for sure 1 running instance, that's why I was mentioning that at the `FoldingScheme` trait level we [already have it](https://github.com/privacy-scaling-explorations/folding-schemes/tree/b2c2f3fb7b24c4ec75d186b036aecdb6fc21ec23/src/lib.rs#L53) with 1 single running instance.
But here this is for the NIMFS as a multifolding scheme, where I was trying to keep it more close to the scheme described in the HyperNova paper allowing to input multiple LCCCS & multiple CCCS instances, having also the `NIMFS` structure stateless, similar to how the Nova [is done](
https://github.com/privacy-scaling-explorations/folding-schemes/tree/b2c2f3fb7b24c4ec75d186b036aecdb6fc21ec23/src/folding/nova/nifs.rs#L25), and also the [microsoft/Nova/nifs](https://github.com/microsoft/Nova/blob/main/src/nifs.rs#L39), where in both cases, the running instance (previous folds) and the incomming instance are passed as inputs, not as state of the NIFS object. And the restriction to a single LCCCS instance is then done at the IVC/FoldingScheme level as designed in the `FoldingScheme` trait.

But I don't have a big opposition, so if after seeing the separation between the NIMFS and the IVC/FoldingScheme level you still thing that is worth to restrict the `NIMFS` logic to a single instance and also to put that instance as a parameter kind of the 'state' of the NIMFS object, I'm happy to update it ^^

Reply:
Update: we discussed it IRL with @CPerezz , and we agree to keep it as it is for the moment.
Some of the reasons to keep it 'stateless' are:
- Nova's NIFS impl also gets the instances as input parameters, not into the state of the NIFS struct:  https://github.com/privacy-scaling-explorations/folding-schemes/tree/b2c2f3fb7b24c4ec75d186b036aecdb6fc21ec23/src/folding/nova/nifs.rs#L25
- microsoft/Nova/nifs also gets the instances as input parameters, not into the state of the struct: https://github.com/microsoft/Nova/blob/main/src/nifs.rs#L39
- the trait that we have of the FoldingScheme also gets it as inputs of the method prove: https://github.com/privacy-scaling-explorations/folding-schemes/tree/b2c2f3fb7b24c4ec75d186b036aecdb6fc21ec23/src/lib.rs#L53
- having NIMFS with the instance as a parameter in the struct would make it be a different approach than the approach that we have for Nova's NIFS, and when implementing the FoldingScheme trait for HyperNova, would mean to build the struct in each step which holds the 'state' (the instance) inside it, to be lost at the end of the folding step, needing to construct it again in the next step (to fit into the [FoldingScheme](https://github.com/privacy-scaling-explorations/folding-schemes/tree/b2c2f3fb7b24c4ec75d186b036aecdb6fc21ec23/src/lib.rs#L53) trait).

Anyways, once we start to use this from higher levels of the library, we will see more from the practical side if makes more sense to put things differently, which would be a very simple change (but if we change it we would not only change it in HyperNova but also on Nova and in the FoldingScheme trait).